src/Compilers/CSharp/Portable/Emitter/Model/SynthesizedPrivateImplementationDetailsStaticConstructor.cs

```diff
@@ -36,13 +36,14 @@ internal override void GenerateMethodBody(TypeCompilationState compilationState,
             // index portion of the greatest method definition token in the compilation. This guarantees that any
             // method can use the index portion of its own method definition token as an index into the payload array.
 
-            IReadOnlyCollection<KeyValuePair<int, InstrumentationPayloadRootField>> payloadRootFields = ContainingPrivateImplementationDetailsType.GetInstrumentationPayloadRoots();
-            Debug.Assert(payloadRootFields.Count > 0);
+            IOrderedEnumerable<KeyValuePair<int, InstrumentationPayloadRootField>> payloadRootFields = ContainingPrivateImplementationDetailsType.GetInstrumentationPayloadRoots();
+            int payloadRootFieldsCount = payloadRootFields.Count();
+            Debug.Assert(payloadRootFieldsCount > 0);
 
-            ArrayBuilder<BoundStatement> body = ArrayBuilder<BoundStatement>.GetInstance(2 + payloadRootFields.Count);
+            ArrayBuilder<BoundStatement> body = ArrayBuilder<BoundStatement>.GetInstance(2 + payloadRootFieldsCount);
             try
             {
-                foreach (KeyValuePair<int, InstrumentationPayloadRootField> payloadRoot in payloadRootFields.OrderBy(analysis => analysis.Key))
+                foreach (KeyValuePair<int, InstrumentationPayloadRootField> payloadRoot in payloadRootFields)
```

Comment:
The `Count()` and `foreach` will likely both cause `OrderBy` to be evaluated. To avoid that, perhaps drop the call to `Count()` (so no size parameter to `GetInstance()`), or have `GetInstrumentationPayloadRoots` return an `ImmutableArray<KeyValuePair>`, or revert to your previous approach of calling `OrderBy` here. (Sorry for generating the extra work.)